src/ai/ai.cpp

```diff
@@ -4191,6 +4189,23 @@ void assign_targets(sys::state& state, dcon::nation_id n) {
 			}
 		}
 	}
+	/* Our allies (mainly our substates, vassals) - we need to care of them! */
+	std::vector<dcon::nation_id> allied_with;
+	allied_with.reserve(state.world.nation_size());
+	for(const auto ovr : state.world.nation_get_overlord_as_ruler(n))
+		allied_with.push_back(ovr.get_subject());
+	for(auto w : allied_with) {
```

Comment:
rather than iterating to fill the vector and then iterating over its contents, you could merge these two loops and get rid of the vector